orc/src/main/java/org/apache/iceberg/orc/ORCSchemaUtil.java

```diff
@@ -0,0 +1,488 @@
+    List<TypeDescription> children = orcSchema.getChildren();
+    List<String> childrenNames = orcSchema.getFieldNames();
+    Preconditions.checkState(children.size() == childrenNames.size(),
+        "Error in ORC file, children fields and names do not match.");
+
+    List<Types.NestedField> icebergFields = Lists.newArrayListWithExpectedSize(children.size());
+    AtomicInteger lastColumnId = new AtomicInteger(orcSchema.getMaximumId());
+    for (int i = 0; i < children.size(); i++) {
+      icebergFields.add(convertOrcToIceberg(children.get(i), childrenNames.get(i),
+          lastColumnId::incrementAndGet));
+    }
+
+    return new Schema(icebergFields);
+  }
+
+  /**
+   * Converts an Iceberg schema to a corresponding ORC schema within the context of an existing
+   * ORC file schema.
+   * This method also handles schema evolution from the original ORC file schema
+   * to the given Iceberg schema. It builds the desired reader schema with the schema
+   * evolution rules and pass that down to the ORC reader,
+   * which would then use its schema evolution to map that to the writer’s schema.
+   *
+   * Example:
+   * <code>
+   * Iceberg writer                                        ORC writer
+   * struct&lt;a (1): int, b (2): string&gt;                     struct&lt;a: int, b: string&gt;
+   * struct&lt;a (1): struct&lt;b (2): string, c (3): date&gt;&gt;     struct&lt;a: struct&lt;b:string, c:date&gt;&gt;
+   * </code>
+   *
+   * Iceberg reader                                        ORC reader
+   * <code>
+   * struct&lt;a (2): string, c (3): date&gt;                    struct&lt;b: string, c: date&gt;
+   * struct&lt;aa (1): struct&lt;cc (3): date, bb (2): string&gt;&gt;  struct&lt;a: struct&lt;c:date, b:string&gt;&gt;
+   * </code>
+   *
+   * @param schema an Iceberg schema
+   * @param originalOrcSchema an existing ORC file schema
+   * @return the resulting ORC schema
+   */
+  public static TypeDescription buildOrcProjection(Schema schema,
+                                                   TypeDescription originalOrcSchema) {
+    AtomicInteger lastColumnId = new AtomicInteger(getMaxIcebergId(originalOrcSchema));
+    final Map<Integer, OrcField> icebergToOrc = icebergToOrcMapping("root",
+        originalOrcSchema, lastColumnId::incrementAndGet);
+    return buildOrcProjection(Integer.MIN_VALUE, schema.asStruct(), true, icebergToOrc);
+  }
+
+  private static TypeDescription buildOrcProjection(Integer fieldId, Type type, boolean isRequired,
+                                                    Map<Integer, OrcField> mapping) {
+    final TypeDescription orcType;
+
+    switch (type.typeId()) {
+      case STRUCT:
+        orcType = TypeDescription.createStruct();
+        for (Types.NestedField nestedField : type.asStructType().fields()) {
+          String name = Optional.ofNullable(mapping.get(nestedField.fieldId()))
+              .map(OrcField::name)
+              .orElse(nestedField.name());
```

Comment:
We should probably add a unique suffix to the field name in case it is not found in the mapping like in https://github.com/apache/incubator-iceberg/blob/master/core/src/main/java/org/apache/iceberg/avro/BuildAvroProjection.java#L100. This is helpful to avoid issues if the file schema has a field with the same name as a field in the expected schema but at a different id.

Reply:
> This is helpful to avoid issues if the file schema has a field with the same name as a field in the expected schema but at a different id.
I'm not following - I may be missing something but, if the ORC file schema has a field with the same name as a field in the expected (Iceberg) schema should be irrelevant in this case since they have different ids - as we match by id - no?

Reply:
Consider this case:
A field `d` was renamed to `c` and a new field `d` was added later, which resulted in the following schema
Iceberg Schema: 1:a 2:b 3:c 4:d
Orc File Schema: 1:a 2:b 3:d <- This is what `mapping` variable will contain.

Now when going to through Iceberg schema in line 255 it wont find corresponding mapping for field id 4 in `mapping`. So it will use the nestedField.name() according to line 257 which is `d`. So it will project the value for `d` in ORC file for both Iceberg id 3 and 4 which is incorrect.

Reply:
Thanks for the clear explanation, @shardulm94! - I've added a test for this in `org.apache.iceberg.data.TestReadProjection` since it did not catch this issue before. I've added the suffix as well.